src/toil/test/jobStores/jobStoreTest.py

```diff
@@ -564,6 +690,109 @@ def _batchDeletionSize(self):
+        JobStoreFileID = self.master.importFile(srcUrl)
+        self.assertEqual(self._hashKey(keyName), self._hashJobStoreFileID(JobStoreFileID))
+        return JobStoreFileID
+
+    def _hashFile(self, localFilePath):
+        totalSize = os.path.getsize(localFilePath)
+        with open(localFilePath, 'r') as f:
+            if totalSize <= 2**20 * 5:
+                return hashlib.md5(f.read()).hexdigest()
+
+            partSize = 2**20 * 5
+            partCount = 0
+            curr = 0
+            hashList = []
+            while curr < totalSize:
+                partCount += 1
+                hashList.append(hashlib.md5(f.read(partSize)).digest())
+                curr += partSize
+
+            return "%s-%d" % (hashlib.md5(''.join(hashList)).hexdigest(), partCount)
+
+    def _hashJobStoreFileID(self, JobStoreFileID):
+        bucket = self.master.filesBucket
+        keyName = JobStoreFileID
+        info = self.master.FileInfo.loadOrFail(JobStoreFileID)
+        key = bucket.get_key(keyName, headers=info._s3EncryptionHeaders())
+
+        if key is None and bucket.name == self.master.filesBucket.name and self.master.fileExists(keyName):
+            info = self.master.FileInfo.loadOrFail(keyName)
+            return hashlib.md5(info.content).hexdigest()
+        elif key is None:
+            raise RuntimeError("key '%s' does not exist in JobStore" % keyName)
+
+        return key.etag[1:-1]
+    
+    def _hashKey(self, keyName, bucketName=None):
+        bucket = self.s3.get_bucket(bucketName) if bucketName is not None else self.srcBucket
+        key = bucket.get_key(keyName)
+        if key is None:
+            raise RuntimeError("key '%s' does not exist in bucket '%s'" % (keyName, bucketName))
+        return key.etag[1:-1]
+
+    def __uploadToBucketKeyOfSize(self, file_size, bucketName=None):
+        """
+        Uploads random bytes to a file in the file bucket of the aws jobstore
+        of size file_size. If the file size is 50MiB or more multipart upload
+        is used. The optional parameter bucketName allows a different bucket
+        in s3 to be picked.
+
+        :param file_size int: size of file in bytes
+        :return: a pointer to the key that was uploaded
+        """
+        partSize = 2**20 * 50
+
+        bucket = self.s3.get_bucket(bucketName) if bucketName is not None else self.srcBucket
+        keyName = '%.0fMiB_file_%s' % (float(file_size/(2**20)), str(uuid.uuid1()))
+        key = Key(bucket)
+        key.key = keyName
+        with open('/dev/urandom', 'r') as f:
+            if file_size < 2**20*5:
+                key.set_contents_from_string(f.read(file_size))
+            else:  # Multipart
+                mp = bucket.initiate_multipart_upload(key_name=keyName)
+                start = 0
+                partNum = itertools.count()
+                try:
+                    while start <= file_size:
+                        end = min(start + partSize, file_size)
```

Comment:
because the upload_part_from_file we are specifying size, its irrelevant now anyway because of Hannes's earlier suggestion to treat end as exclusive until the last moment.